python/wikibot/extract_dates.py

```diff
@@ -0,0 +1,149 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Class for extracting birth date facts and storing them in a record file"""
+
+import sling
+import re
+import sys
+import json
+
+class ExtractDates:
+  def __init__(self):
+    self.kb = sling.Store()
+    self.kb.load("local/data/e/wiki/kb.sling")
+    self.instanceof = self.kb['P31']
+    self.has_part = self.kb['P527']
+    self.part_of = self.kb['P361']
+    self.item_category = self.kb['/w/item/category']
+    self.date_of_birth = self.kb['P569']
+    self.wikimedia_category = self.kb['Q4167836']
+    self.date_types = [
+      self.kb['Q29964144'], # year BC
+      self.kb['Q577'],      # year
+      self.kb['Q39911'],    # decade
+      self.kb['Q578'],      # century
+      self.kb['Q36507'],    # millennium
+    ]
+    self.human = self.kb['Q5']
+
+    self.calendar = sling.Calendar(self.kb)
+    self.names = sling.PhraseTable(self.kb,
+                                   "local/data/e/wiki/en/phrase-table.repo")
+    self.kb.freeze()
+    self.date_type = {}
+    self.conflicts = 0
+    self.out_file = "local/data/e/wikibot/birth_dates.rec"
+
+  def find_date(self, phrase):
+    for item in self.names.lookup(phrase):
+      for cls in item(self.instanceof):
+        clsr = cls.resolve()
+        if clsr in self.date_types:
+          self.date_type[item] = clsr
+          return item
+    return None
+
+  def dated_categories(self, pattern):
+    cats = {}
+    pat = re.compile(pattern)
+
+    for item in self.kb:
+      is_category = False
+      for cls in item(self.instanceof):
+        if cls == self.wikimedia_category: is_category = True
+      if not is_category: continue
+      name = str(item.name)
+      m = pat.match(name)  # , re.IGNORECASE
+      if m is not None:
+        date = self.find_date(m.group(1))
+        if date is not None:
+          cats[item] = date
+    print "Done with dated categories", pattern, len(cats)
+    return cats
+
+  def most_specific_date(self, dates):
+    if dates is None: return None
+    if len(dates) == 1: return dates[0]
+    dt = {}
+    for cat, date in dates:
+      dti = self.date_type[date]
+      if not dti: return None # no date type found for the date
+      if dti in dt:
+        self.conflicts += 1
+        return None # conflict from two dates of the same date type
+      dt[dti] = (cat, date)
+    # TODO: check that there is consistency across date types, e.g., can't have
+    # both born in the year 1911 and in the decade the 1920s
+    for i in range(len(self.date_types)):
+      if self.date_types[i] in dt:
+        return dt[self.date_types[i]]
+    return None
+
+  def find_births(self):
+    globals = sling.Store()
```

Comment:
It might be easier to add the names below to the store initialized in the constructor instead of making a new globals store here.

Reply:
Yes. That simplifies things quite a bit. Fixed.